core/app/models/spree/store_product.rb

```diff
@@ -5,6 +5,8 @@ class StoreProduct < Spree::Base
     belongs_to :store, class_name: 'Spree::Store', touch: true
     belongs_to :product, class_name: 'Spree::Product', touch: true
 
+    after_destroy -> { product.destroy! if Spree::StoreProduct.where.not(store_id: store.id).find_by(product_id: product.id).nil? }
```

Comment:
```suggestion
    after_destroy -> { product.destroy! unless self.class.where(product_id: product.id).not(store_id: store.id).exists? }
```

Reply:
@tomurb also let's move this logic into a method - `destroy_associated_product`

so the code would be

```ruby
after_destroy :destroy_associated_product
```

```ruby
protected

def destroy_associated_product
  return if self.class.where(product_id: product.id).not(store_id: store.id).any?
  
  product.destroy!
end
```